compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -164,5 +165,31 @@ pub(crate) fn deref_mir_constant<'tcx>(
         },
     };
 
-    mir::ConstantKind::Val(op_to_const(&ecx, &mplace.into()), ty)
+    mir::ConstantKind::Val(op_to_const(&mut ecx, &mplace.into()), ty)
+}
+
+pub fn slice_for_alloc<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    alloc: ConstAllocation<'tcx>,
+) -> ConstAllocation<'tcx> {
+    let addr = tcx.create_memory_alloc(alloc);
+    let ptr_size = tcx.data_layout.pointer_size;
+
+    let a = Scalar::Ptr(Pointer::new(addr, Size::ZERO), ptr_size.bytes().try_into().unwrap());
+    let b = Scalar::Int(ty::ScalarInt::try_from_uint(alloc.0.size().bytes(), ptr_size).unwrap());
+    let imm = crate::interpret::Immediate::ScalarPair(a, b);
+    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, ty::ParamEnv::reveal_all(), false);
+
+    let layout = tcx
+        .layout_of(
+            ty::ParamEnv::reveal_all()
+                .and(tcx.mk_imm_ref(tcx.lifetimes.re_static, tcx.mk_slice(tcx.types.u8))),
+        )
+        .unwrap();
+    let place = ecx.allocate(layout, crate::interpret::MemoryKind::Stack).unwrap();
+    ecx.write_immediate(imm, &place.into()).unwrap();
+    intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &place).unwrap();
+
+    assert_eq!(place.ptr.into_parts().1, Size::ZERO);
+    tcx.global_alloc(place.ptr.provenance.unwrap()).unwrap_memory()
```

Comment:
Ah, I was hoping we could avoid this duplication... can this call `op_to_const` or would that be a bad idea? It'd have to return a `ConstValue` instead of `ConstAllocation` but that might be good enough for all callers?

Reply:
Can't call that as we have no OpTy and we can't build one from scratch without breaking abstractions

Reply:
You can build an `ImmTy`, and then `into()` than into an `OpTy`.

Reply:
I cannot build an `ImmTy`, its `imm` field is also private.

Reply:
`ImmTy::from_immediate` is public